### PR TITLE
refactor(ir): support variadic arguments, typehint annotations and add generic graph traversal routines

### PR DESCRIPTION
**Pull request by kszucs**
- allow custom fields for Annotable
- support type hints instead of validators
- reorganize signature merging to annotation.py
- introduce traversable "trait" 
- add generic graph traversal routines

### CODE REVIEW

File: ibis/common/grounds.py
Comment (cpcloud) [inline ibis/common/grounds.py:240]: of the ... ? 😂

Comment (kszucs) [reply ibis/common/grounds.py]: fixed 😄

File: ibis/backends/base/sql/__init__.py
Comment (cpcloud) [inline ibis/backends/base/sql/__init__.py:204]: Can we remove `_find_memtables` or replace it with `lambda op: (graph.proceed, op)` and pass `types=ops.InMemoryTable`?

File: ibis/common/annotations.py
Comment (cpcloud) [inline ibis/common/annotations.py:49]:
```suggestion
        return f"{self.__class__.__name__}({self.validator!r}, default={self.default!r})"
```

File: ibis/common/annotations.py
Comment (cpcloud) [inline ibis/common/annotations.py:81]:
```suggestion
        return f"{self.__class__.__name__}({self.validator!r})"
```

File: ibis/common/annotations.py
Comment (cpcloud) [inline ibis/common/annotations.py:143]: Might be useful to give a quick blurb about what `this` is.

File: ibis/common/graph.py
Comment (cpcloud) [inline ibis/common/graph.py:28]: Can we call this `from_bfs`/`from_dfs`? It feels like `dfs` and `bfs` would be instance methods if anything.

File: ibis/common/tests/test_grounds.py
Comment (cpcloud) [inline ibis/common/tests/test_grounds.py:106]: You've just made ibis into a Lisp 😂

File: ibis/expr/visualize.py
Comment (cpcloud) [inline ibis/expr/visualize.py:84]: I think I'd prefer the slightly-more-verbose `Graph.from_bfs(expr)` here, though I could also see a `Graph.from_expr(expr, algo=traverse.bfs)` 🤷🏻 

Not super opinionated here.

File: ibis/util.py
Comment (cpcloud) [inline ibis/util.py:74]: I'm surprised we managed to avoid this pattern for this long!

**Review by cpcloud — APPROVED** [review]
LGTM!

File: ibis/common/grounds.py
Comment (kszucs) [inline ibis/common/grounds.py:267]: Cover it with tests

File: ibis/common/annotations.py
Comment (kszucs) [reply ibis/common/annotations.py:81]: Updated.

File: ibis/expr/visualize.py
Comment (kszucs) [reply ibis/expr/visualize.py:84]: Updated to `Graph.from_bfs(expr)`

File: ibis/common/graph.py
Comment (kszucs) [reply ibis/common/graph.py:28]: Updated.

**Comment by cpcloud** [comment]
@kszucs Can you squash? Will merge after that!

File: ibis/common/grounds.py
Comment (kszucs) [reply ibis/common/grounds.py]: Implementing traversable for ops.Node instead.

**Comment by kszucs** [comment]
@cpcloud it requires another review

**Comment by cpcloud** [comment]
> @cpcloud it requires another review

Got it.

What's the breaking change here? APIs in ibis.util are not guaranteed to be backwards compatible even if they are public.

**Comment by kszucs** [comment]
> > @cpcloud it requires another review
> 
> Got it.
> 
> What's the breaking change here? APIs in ibis.util are not guaranteed to be backwards compatible even if they are public.

I removed the breaking change note from the commit then.

**Comment by kszucs** [comment]
@cpcloud it's green

File: ibis/expr/operations/reductions.py
Comment (kszucs) [inline ibis/expr/operations/reductions.py:109]: Perhaps we should just call it `initialize`?

Comment (cpcloud) [reply ibis/expr/operations/reductions.py:109]: `intialized` is fine for now IMO

**Review by cpcloud — APPROVED** [review]
LGTM! You can ignore the docstring comments.

File: ibis/backends/clickhouse/__init__.py
Comment (kszucs) [reply ibis/backends/clickhouse/__init__.py:44]: Yes

File: ibis/common/caching.py
Comment (kszucs) [reply ibis/common/caching.py:72]: Not really, trying to switch to it then.

File: ibis/backends/base/sql/compiler/select_builder.py
Comment (kszucs) [reply ibis/backends/base/sql/compiler/select_builder.py:42]: `Annotable.__argnames__` contains the attribute names.

`Concrete.__args__` contains the attribute values.
`Concrete.args` is a backward compatible alias for `__args__`.
`Concrete.argnames` is a backward compatible for `__argnames__`.

Since `NodeList.values` is a tuple of nodes, `NodeList.args == ((values...),)` contains another level of nesting. This is where `__children__` comes into the picture since `Signature.unbind()` unflattens variadic arguments. 

In the future we could port the SQL compiler to use the graph traversal functions which would make this line obsolete.

Comment (cpcloud) [reply ibis/backends/base/sql/compiler/select_builder.py:42]: Got it, sounds good. We'll get to that port eventually!

File: ibis/common/annotations.py
Comment (kszucs) [reply ibis/common/annotations.py:26]: How about configuring a doc formatter to automatically handle these? Like https://github.com/PyCQA/docformatter if black doesn't support docstring formatting.

File: pyproject.toml
Comment (kszucs) [reply pyproject.toml:39]: Since the type annotations required at runtime we need this as a mandatory dependency.

Comment (cpcloud) [reply pyproject.toml:39]: Got it! Thanks.

File: ibis/common/caching.py
Comment (kszucs) [reply ibis/common/caching.py]: Done

**Review by cpcloud — APPROVED** [review]

**Pull request merged**
